File: src/java/server/com/topvision/ems/fault/parser/trapParser.rb
# coding: UTF-8
require 'java'

java_import 'com.topvision.ems.fault.parser.TrapParser'
java_import 'com.topvision.ems.fault.domain.Event'
java_import 'com.topvision.ems.fault.service.EventSender'
java_import 'java.util.Iterator'
class RubyTrapParser
    def getTrapCos()
        return 10000;
    end
	def parse(entityId,trap)
		@@trap = trap
		begin
			uptime = trap.getVarialbleValue("1.3.6.1.2.1.1.3.0")
			type = trap.getVarialbleValue("1.3.6.1.6.3.1.1.4.1.0")
			event = EventSender.getInstance().createEvent(1,trap.getAddress(),trap.getVarialbleValue("1.3.6.1.6.3.18.1.3.0"))
			# coldStart
			if(type == "1.3.6.1.6.3.1.1.5.1")
				event.setTypeId(10004)
				#event.setMessage("设备冷启动")
				event.setMessage("\350\256\276\345\244\207\345\206\267\345\220\257\345\212\250")
			# warmStart
			elsif(type == "1.3.6.1.6.3.1.1.5.2")
				event.setTypeId(10005)
				#event.setMessage("设备热启动")
				event.setMessage("\350\256\276\345\244\207\347\203\255\345\220\257\345\212\250")
			# linkDown
			elsif(type == "1.3.6.1.6.3.1.1.5.3")
				event.setTypeId(30003)
				trap.getVariableBindings().keySet.each{|key|
				    key =~ /1\.3\.6\.1\.2\.1\.2\.2\.1\.1\./
				    if($' != nil)
				        @ifIndex = trap.getVariableBindings().get(key).to_i
				    end
				}
				event.setSource("#{@ifIndex}")
				#event.setMessage("#{@ifIndex}端口Down")
				event.setMessage("#{@ifIndex}\347\253\257\345\217\243Down")
			# linkUp
			elsif(type == "1.3.6.1.6.3.1.1.5.4")
				event.setTypeId(30004)
				trap.getVariableBindings().keySet.each{|key|
				    key =~ /1\.3\.6\.1\.2\.1\.2\.2\.1\.1\./
				    if($' != nil)
				        @ifIndex = trap.getVariableBindings().get(key).to_i
				    end
				}
				event.setSource("#{@ifIndex}")
				#event.setMessage("#{@ifIndex}端口Up")
				event.setMessage("#{@ifIndex}\347\253\257\345\217\243Up")
				event.setClear(true)
			# authenticationFailure
			elsif(type == "1.3.6.1.6.3.1.1.5.5")
				event.setTypeId(60003)
				#event.setMessage("登录认证失败")
				event.setMessage("\347\231\273\345\275\225\350\256\244\350\257\201\345\244\261\350\264\245")
			else
				#puts '>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>'
				#puts trap.to_s
				#puts '>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>'
				event = nil
			end
			if(event != nil)
				EventSender.getInstance().send(event)
				return false
			end
		rescue
		    puts "#{$!}"
		    puts "#{$@}"
		end
		return true
	end
end

File: src/java/server/com/topvision/ems/fault/parser/TrapParser.java
/***********************************************************************
 * $Id: SyslogParser.java,v1.0 2012-1-13 下午01:38:41 $
 * 
 * @author: Victor
 * 
 * (c)Copyright 2011 Topvision All rights reserved.
 ***********************************************************************/
package com.topvision.ems.fault.parser;

import com.topvision.ems.fault.domain.Event;
import com.topvision.ems.fault.domain.EventSource;
import com.topvision.ems.fault.exception.TrapInstanceException;
import com.topvision.framework.snmp.Trap;

/**
 * @author Victor
 * @created @2012-1-13-下午01:38:41
 * 
 *          Trap解析接口，实现此接口并把实现注册到TrapService中，接收到Trap会顺序解析
 */
public interface TrapParser {

    /**
     * 判断trap是否匹配当前trapParser
     * 
     * @param entityId
     * @param trap
     * @return
     */
    public boolean match(Long entityId, Trap trap);

    /**
     * 解析告警，所有trap的解析都按照规定的5个步骤的模版来执行，各子类只需实现其中的五步即可</br>
     * 1. parseTrapCode 解析trapCode </br>
     * 2. convertToEmsCode 获取网管对应的code </br>
     * 3. parseEventSource 根据trap解析对应的位置</br>
     * 4. buildEvent 根据trap构造event </br>
     * 5. sendEvent 发送转换后的事件
     * 
     * @param entityId
     *            来源设备ID
     * @param trap
     *            接收的Trap
     */
    public void parse(Long entityId, Trap trap);

    /**
     * 解析trapCode
     * 
     * @param trap
     * @return
     */
    public Long parseTrapCode(Trap trap);

    /**
     * 根据trap获取网管对应的code
     * 
     * @param trap
     *            接收到的trap
     * @param trapCode
     *            trap的code
     * @return
     */
    public Integer convertToEmsCode(Long entityId, Trap trap, Long trapCode);

    /**
     * 根据trap解析对应的位置
     * 
     * @param trap
     *            接收到的trap
     * @param entityId
     *            entityId
     * @param trapCode
     *            trap的code
     * @param emsCode
     *            网管对应的code
     * @return
     * @throws TrapInstanceException
     */
    public EventSource parseEventSource(Trap trap, Long entityId, Long trapCode, Integer emsCode)
            throws TrapInstanceException;

    /**
     * 根据trap构造event
     * 
     * @param trap
     *            接收到的trap
     * @param trapCode
     *            trap的code
     * @param emsCode
     *            网管对应的code
     * @param eventSource
     *            trap对应设备上的位置
     * @return
     */
    public Event buildEvent(Trap trap, Long trapCode, Integer emsCode, EventSource eventSource);

    /**
     * 发送转换后的事件
     * 
     * @param trap
     * @param event
     */
    public void sendEvent(Trap trap, Event event);

    /**
     * 获取该trapParser对应的优先级
     * 
     * @return
     */
    public Integer getTrapCos();
}
